doc: Add information on Twister Statuses (#80625)
zephyrproject-rtos/zephyr: doc/develop/test/twister_statuses.rst

.. _twister_statuses:

Twister Status
##############

What is a Twister Status?
=========================

Twister Status formulates the current state of

- ``Harness``
- ``TestCase``
- ``TestSuite``
- ``TestInstance``

in a comprehensive and easy-to understand way.
In practice, most users will be interested in Statuses
of Instances and Cases after the conclusion of their Twister runs.

.. tip::

   Nomenclature reminder:

   .. tabs::

      .. tab:: ``Harness``

         ``Harness`` is a Python class inside Twister that allows us
         to capture and analyse output from a program external to Twister.

      .. tab:: ``TestCase``

         ``TestCase`` is a piece of code that aims to verify some assertion.
         It is the smallest subdivision of testing in Zephyr.

      .. tab:: ``TestSuite``

         ``TestSuite`` is a grouping of Test Cases.
         One can modify Twister's behaviour on a per-Suite basis via ``testcase.yaml`` files.
         Such grouped Cases should have enough in common
         for it to make sense to treat them all the same by Twister.

      .. tab:: ``TestInstance``

         ``TestInstance`` is a Test Suite on some platform.

>          Twister reports typically report the results for Instances,

Author: Changed to ` Twister typically reports its results for Instances,`

         despite them being called "Suites" there.
         As the distinction between them is not useful in this section,
         whenever you read "Suite", assume the same for Instances.

   More detailed explanation can be found :ref:`here <twister_tests_long_version>`.

Possible Twister Statuses
=========================

.. list-table:: Twister Statuses
   :widths: 10 10 70 10
   :header-rows: 1

   * - In-code
     - In-text
     - Description

>      - Where applicable

Author: FILTER -> NOTRUN -> BLOCK -> SKIP -> ERROR -> FAIL -> PASS would be a better order, with the not executed statuses appearing first, following the general Twister pipeline and with the PASS placed last.

   * - NOTRUN
     - not run
     - The test wasn't run because it was not runnable in current configuration.
       It was, however, built correctly.

>      - Suite, Case

Author: Added both names to the Tip section and changed all `Test Suite`, `Test Case`, etc. into `Suite`, `Case`, etc., respectively.

   * - BLOCK
     - blocked
     - The test was not run because of an error or crash in the test suite.
     - Case
   * - FILTER
     - filtered
     - A static or runtime filter has eliminated the test from the list of tests to use.
     - Suite, Case
   * - ERROR
     - error
     - The test produced an error in running the test itself.
     - Suite, Case
   * - FAIL
     - failed
     - The test produced results different than expected.
     - Suite, Case
   * - PASS
     - passed
     - The test produced results as expected.
     - Suite, Case
   * - SKIP
     - skipped
     - Test was skipped by some other reason than previously delineated.
     - Suite, Case

**In-code** and **In-text** columns should the names of a given status.
If one is not a Twister developer, the in-text name should suffice.
It is the name used in the JSON reports.

**Where applicable** column indicates where we might find such Statuses.

> Harness has been omitted for clarity, as it does not appear in final report files.

Reviewer: these two sentences, esp. for Harness, might be moved to the 'Nomenclature reminder' module.

Author: Makes sense, especially with the table column change.

If a status is marked as applicable for Suites, it is also applicable for Instances.

.. note::

   There are two more Statuses that are internal to Twister.

>    ``NONE`` (A starting status for Cases and Suites) and
>    ``STARTED`` (Indicating an in-progress Case).

Reviewer: with bullets
```suggestion

   -  ``NONE`` - an initial status for Cases and Suites,
   - ``STARTED`` - Indicating an in-progress Case.
   
```

Author: Done, will have to see how it interacts with the Note box.

Author: Doesn't play nicely with the box, had to remove it.

Reviewer: note the proposed changes in the text itself.

   Those should not appear in the final Twister report.
   Appearance of non-terminal Statuses in one's report files indicates a problem with Twister.

Test Case and Test Suite Status combinations
============================================

.. list-table:: Test Case and Test Suite Status combinations
   :widths: 22 13 13 13 13 13 13
   :align: center
   :header-rows: 1
   :stub-columns: 1

>    * - Case\\Suite

Reviewer: ```suggestion
   * - TestCase\\TestSuite
```
would be nice to have some arrows here (down and right)

     - FILTER
     - ERROR
     - FAIL
     - PASS
     - NOTRUN
     - SKIP
   * - FILTER
     - ✓
     - ✕
     - ✕
     - ✕
     - ✕
     - ✕
   * - ERROR
     - ✕
     - ✓
     - ✕
     - ✕
     - ✕
     - ✕
   * - BLOCK
     - ✕
     - ✓
     - ✓
     - ✕
     - ✕
     - ✕
   * - FAIL
     - ✕
     - ✓
     - ✓
     - ✕
     - ✕
     - ✕
   * - PASS
     - ✕
     - ✓
     - ✓
     - ✓
     - ✕
     - ✕
   * - NOTRUN
     - ✕
     - ✓
     - ✓
     - ✓
     - ✓
     - ✕
   * - SKIP
     - ✕
     - ✓
     - ✓
     - ✓
     - ✓
     - ✓

✕ indicates that such a combination should not happen in a proper Twister run. In other words,
no Suite of a status indicated by the table column should contain any Cases of a status indicated

> by the table row. ✓ indicates a proper combination.

Reviewer: ```suggestion
by the table row.

  ✓ indicates a proper combination.
```

> Detailed explanation, per Test Suite Status

Reviewer: Isn't this paragraph almost the same as what is in the 'Possible Twister Statuses' table and can be moved there ?
If a state meaning is different for TestSuite vs. TestCase they can be split into table columns.

Author: While the first sentence, the general Status description, is very similar to the table, in this section I aimed to explain why some combinations are possible and others are not.
When a documentation reader encounters the combination table, I'd assume they'd like to ask *why* certain combinations are impossible. So this needs to be after the combination table.

Reviewer: unfortunately, having two descriptions tend to diverge from each other not clarifying much,
for example `FILTER`:
* `A static or runtime filter has eliminated the test from the list of tests to use.`
* `This status indicates that the whole Suite has been statically filtered out of a given Twister run. Thus, any Case within it should also have such a status.`
The second sentence don't mention runtime filtering and it is still vague why other combinations are impossible (if it is the goal to answer here), e.g. together with `PASS` why Suite: PASS and Case:FILTER is impossible ? same for `NOTRUN`.

`SKIP` is another example as it explains only one combination, isn't it ?

.. _twister_statuses:

Twister Status
##############

What is a Twister Status?
=========================

Twister Status formulates the current state of

- ``Harness``
- ``TestCase``
- ``TestSuite``
- ``TestInstance``

in a comprehensive and easy-to understand way.

> In practice, most users will be interested in Statuses
> of Instances and Cases after the conclusion of their Twister runs.

Reviewer: ```suggestion
```
IMO its better to remove this from here to allow more clarification of what 'instance' is from user's perspective later in the 'nomenclature' section below.

   .. tabs::

      .. tab:: ``Harness``

         ``Harness`` is a Python class inside Twister that allows us
         to capture and analyse output from a program external to Twister.
         It has been excised from this page for clarity,
         as it does not appear in final reports.

      .. tab:: ``TestCase``

         ``TestCase``, also called Case, is a piece of code that aims to verify some assertion.
         It is the smallest subdivision of testing in Zephyr.

      .. tab:: ``TestSuite``

         ``TestSuite``, also called Suite, is a grouping of Cases.
         One can modify Twister's behaviour on a per-Suite basis via ``testcase.yaml`` files.
         Such grouped Cases should have enough in common
         for it to make sense to treat them all the same by Twister.

      .. tab:: ``TestInstance``

         ``TestInstance``, also called Instance, is a Suite on some platform.
         Twister typically reports its results for Instances,
         despite them being called "Suites" there.
         If a status is marked as applicable for Suites, it is also applicable for Instances.
         As the distinction between them is not useful in this section,

>          whenever you read "Suite", assume the same for Instances.

Reviewer: Twister summary output uses `test configurations executed on platforms` for 'instances'

   More detailed explanation can be found :ref:`here <twister_tests_long_version>`.

Possible Twister Statuses
=========================

.. list-table:: Twister Statuses
   :widths: 10 10 66 7 7
   :header-rows: 1

   * - In-code

>      - In-text

Reviewer: ```suggestion
     - In-report
```

If a status is marked as applicable for Suites, it is also applicable for Instances.

Test Case and Test Suite Status combinations
============================================

.. list-table:: Test Case and Test Suite Status combinations
   :widths: 16 12 12 12 12 12 12 12
   :align: center
   :header-rows: 1
   :stub-columns: 1

   * - Case\\Suite
     - NONE
     - FILTER
     - ERROR
     - FAIL
     - PASS
     - NOTRUN
     - SKIP
   * - NONE
     - N/A
     - N/A
     - N/A
     - N/A
     - N/A
     - N/A
     - N/A
   * - STARTED
     - N/A
     - N/A
     - N/A
     - N/A
     - N/A
     - N/A
     - N/A
   * - FILTER
     - N/A
     - ✓
     - ✕
     - ✕
     - ✕
     - ✕
     - ✕
   * - ERROR
     - N/A
     - ✕
     - ✓
     - ✕
     - ✕
     - ✕
     - ✕
   * - BLOCK
     - N/A
     - ✕
     - ✓
     - ✓
     - ✕
     - ✕
     - ✕
   * - FAIL
     - N/A
     - ✕
     - ✓
     - ✓
     - ✕
     - ✕
     - ✕
   * - PASS
     - N/A
     - ✕
     - ✓
     - ✓
     - ✓
     - ✕
     - ✕
   * - NOTRUN
     - N/A
     - ✕
     - ✓
     - ✓
     - ✓
     - ✓
     - ✕
   * - SKIP
     - N/A
     - ✕
     - ✓
     - ✓
     - ✓
     - ✓
     - ✓

> ✕ indicates that such a combination should not happen in a proper Twister run. In other words,

Reviewer: What is the meaning for 'N/A's in this table ?
IMO all of them could be excluded from here for simplicity, so the table might keep only 'V'-s where applicable.

Author: As mentioned above and below, those statuses are not terminal, so they should not appear at all. So N/A - not applicable.

This style of checkmarks, crosses and N/As is consistent with other tables - e.g. https://docs.zephyrproject.org/latest/services/device_mgmt/mcumgr.html

Reviewer: > As mentioned above and below, those statuses are not terminal, so they should not appear at all. So N/A - not applicable.

These 'non-terminal' status' combinations are rather Twister internals and having them here filling the table with N/As is confusing, so it would be pretty enough just to have the note below that `.. their existence at the end of a Twister run indicates a problem with the Twister itself.`

Author: Added that information as a note and removed it from the tables.
